### ISSUE TITLE
clang-cl mismangles C++17 `auto` NTTP of pointer-to-member type

### ISSUE THREAD

**Issue by martinbonaciugome** [llvm/llvm-project#70899]
code is quite large, but I get the error:
```
C:\Projects\chimera_client\clientsdk\igl\headers\igl\LinkedList.hpp(212,5): error : definition with same mangled name '??1?$LinkedList@$07@iugo@@QEAA@XZ' as another definition
    212 |     ~LinkedList()
        |     ^
  C:\Projects\chimera_client\clientsdk\igl\headers\igl\LinkedList.hpp(212,5): note: previous definition is here
C:\Projects\chimera_client\clientsdk\igl\headers\igl\LinkedList.hpp(203,5): error : definition with same mangled name '??0?$LinkedList@$07@iugo@@QEAA@$$QEAU01@@Z' as another definition
    203 |     LinkedList(LinkedList&& rhs) noexcept
        |     ^
  C:\Projects\chimera_client\clientsdk\igl\headers\igl\LinkedList.hpp(203,5): note: previous definition is here
```
only on windows with clang-cl

code is, but without includes is:
```cpp
namespace iugo
{
  template<auto NodeMember>
  struct LinkedList;
  
  
#pragma mark - LinkedListNode
  template<class _T>
  struct LinkedListNode
  {
    using T = RemoveReference<_T>;
    using ElementType = Conditional<IsReference<_T>, Ptr<T>, Instance<T>>;

    LinkedListNode() noexcept
    {
    }
    LinkedListNode(LinkedListNode const&) = delete;
    ~LinkedListNode()
    {
    }

    auto& operator =(LinkedListNode const&) = delete;
    
  private:
    template<auto> friend struct LinkedList;

    union
    {
      struct
      {
        Ptr<T> Prev;
        ElementType Next;
      };
    };
#if !FINAL
    bool IsConnected = false;
#endif

    void Create(Ptr<T> prev) noexcept
    {
      iglAssert(!IsConnected);
      unsafe::Construct(Prev, move(prev));
      unsafe::Construct(Next);
#if !FINAL
      IsConnected = true;
#endif
    }
    void Destroy() noexcept
    {
      iglAssert(IsConnected);
      unsafe::Destruct(Next);
      unsafe::Destruct(Prev);
#if !FINAL
      IsConnected = false;
#endif
    }
  };


#pragma mark - LinkedList
  template<class _T, class T, LinkedListNode<_T> T::* NodeMember>
  struct [[clang::trivial_abi]] LinkedList<NodeMember>
  {
    using ElementType = typename LinkedListNode<_T>::ElementType;

#pragma mark - Node
    struct Node: Ptr<T>
    {
      using Ptr<T>::Ptr;
      Node(Ptr<T> rhs) noexcept
      : Ptr<T>{move(rhs)}
      {
      }
      
      using Ptr<T>::operator =;
      
      [[nodiscard]] auto operator ->() const noexcept
      {
        return &Ptr<T>::operator ->*(NodeMember);
      }
    };
    
#pragma mark - Enumerator
    template<class TList>
    struct [[nodiscard]] Enumerator
    {
      TList&& List;

      Enumerator(TList&& list) noexcept
      : List{forward<TList>(list)}
      {
      }

      auto Remove() noexcept requires(!IsConst<RemoveReference<TList>>)
      {
#if !FINAL
        iglAssert(Version == List.Version(), "Modification during enumeration.");
        Version = ++List._Version;

#endif
        return List.Remove(Exchange(Current, (Current->*NodeMember).Next));
      }

      [[nodiscard]] decltype(auto) operator *() const noexcept
      {
        if constexpr(IsRValueReference<TList>)
        {
          return move(*Current);
        }
        else
        {
          return *Current;
        }
      }

      auto& operator ++()
      {
#if !FINAL
        iglAssert(Version == List.Version(), "Modification during enumeration.");

#endif
        Current = (Current->*NodeMember).Next;
        return *this;
      }

      explicit operator bool() const noexcept
      {
        return bool(Current);
      }

    private:
      decltype(declval<TList&>().First()) Current = List.First();
#if !FINAL
      uint Version = List.Version();
#endif
    };

    template<class TList>
    Enumerator(TList&&) -> Enumerator<TList>;

#pragma mark -
    [[nodiscard]] uint Count() const noexcept
    {
      return _Count;
    }
    
    [[nodiscard]] Ptr<T const> First() const noexcept
    {
      return _Head;
    }
    [[nodiscard]] Ptr<T> First() noexcept
    {
      return _Head;
    }
    [[nodiscard]] Ptr<T const> Last() const noexcept
    {
      return _Head ? Node{_Head}->Prev : nullptr;
    }
    [[nodiscard]] Ptr<T> Last() noexcept
    {
      return _Head ? Node{_Head}->Prev : nullptr;
    }

#if !FINAL
    [[nodiscard]] uint Version() const noexcept
    {
      return _Version;
    }
    
#endif
    LinkedList() = default;
    LinkedList(nullptr_t) noexcept
    {
    }
    template<Enumerable TEnum> requires(!IsSame<Unqualify<TEnum>, LinkedList> && (IsInitializable<T, EnumerableType<TEnum>> || IsConvertible<EnumerableType<TEnum>, ElementType>))
    LinkedList(TEnum&& enumerable)
    {
      for (auto&& item : forward<TEnum>(enumerable))
      {
        Add(forward<decltype(item)>(item));
      }
    }
    LinkedList(LinkedList&& rhs) noexcept
    : _Count{rhs._Count}
    , _Head{move(rhs._Head)}
    {
#if !FINAL
      ++rhs._Version;
#endif
      rhs._Count = 0;
    }
    ~LinkedList()
    {
      Destroy(_Head);
    }

    auto& operator =(nullptr_t) noexcept
    {
#if !FINAL
      ++_Version;
#endif
      Destroy(_Head);
      
      _Count = 0;
      unsafe::Reconstruct(_Head);
      return *this;
    }
    auto& operator =(LinkedList&& rhs) noexcept
    {
      iglAssert(this != &rhs);
      
#if !FINAL
      ++_Version;
#endif
      _Count = rhs._Count;
      _Head = move(rhs._Head);

#if !FINAL
      ++rhs._Version;
#endif
      rhs._Count = 0;
      return *this;
    }

    T& Add(ElementType newValue) noexcept
    {
      auto& value = *newValue;

#if !FINAL
      ++_Version;
#endif
      ++_Count;
      
      if (_Head)
      {
        Node node = _Head;
        auto& next = Node{node->Prev}->Next;
        next = move(newValue);
        Node{next}->Create(Exchange(node->Prev, next));
      }
      else
      {
        _Head = move(newValue);
        Node{_Head}->Create(_Head);
      }

      return value;
    }
    template<class... TArgs> requires IsInitializable<T, TArgs...>
    decltype(auto) Add(TArgs&&... args)
    {
      static_assert(IsSame<ElementType, Instance<T>>);
      
      return Add(New<T>(forward<TArgs>(args)...));
    }
    [[nodiscard]] ElementType Extract(Node node) noexcept
    {
      Finally finally = [node] { node->Destroy(); };
      
#if !FINAL
      ++_Version;
#endif
      --_Count;
      
      if (_Head != node)
      {
        Node prev = node->Prev;
        auto& next = prev->Next;
        auto oldValue = Exchange(next, move(node->Next));
        Node{next ? next : _Head}->Prev = prev;
        return oldValue;
      }
      else if (node->Next)
      {
        Node prev = node->Prev;
        auto oldValue = Exchange(_Head, move(node->Next));
        Node{_Head}->Prev = prev;
        return oldValue;
      }
      else
      {
        return move(_Head);
      }
    }
    void Remove(Node node) noexcept
    {
      [[maybe_unused]] ElementType _ = Extract(node);
    }
    [[nodiscard]] ElementType Pop() noexcept
    {
      return Extract(_Head->Prev);
    }

    auto begin()&& noexcept
    {
      return Enumerator{move(*this)};
    }
    auto begin() const& noexcept
    {
      return Enumerator{*this};
    }
    auto begin()& noexcept
    {
      return Enumerator{*this};
    }
    static constexpr auto end() noexcept
    {
      return Sentinel{};
    }
    
    explicit operator bool() const noexcept
    {
      return bool(_Head);
    }

  private:
    uint _Count = 0;
    ElementType _Head;
#if !FINAL
    uint _Version = 0;
#endif
    
    void Destroy(Node node) noexcept
    {
      if (node)
      {
        Destroy(node->Next);

        node->Destroy();
      }
    }
  };
}
```

**Comment by dwblaikie** [comment]
A few comments:
* Seems the example is both long, but incomplete (fails to compile on godbolt with unresolved symbol `RemoveReference`
* comparing clang-cl's behavior in this regard to non-cl compilers (clang/gcc/etc) isn't necessarily valid - compare it to msvc. clang-cl is ABI compatible with MSVC and so has to make different choices that might preclude valid code to ensure compatibility

**Comment by martinbonaciugome** [comment]
Sorry, yes I know it doesn't compile by itself, as it would very hard to do so. Our project is like around 300000 lines of code, and compiles completely except this 1 error. The error is not something that should be possible with a template that is fully contained in a single header. Trying to repro in a single source on godbolt will be very hard for this type of error. I have a workaround, which changes the interface, but is functionally the same is just redundantly adding `class T`:
```
template<class T, auto NodeMember>
struct LinkedList;
template<class _T, class T, LinkedListNode<_T> T::* NodeMember>
struct [[clang::trivial_abi]] LinkedList<T, NodeMember>
```
and with no other changes at all. now to use the class:
```
LinkedList<Object, &Object::Node> _Objects;
```
instead of just:
```
LinkedList<&Object::Node> _Objects;
```
which shouldn't change anything compiler wise, as the specialization of `LinkedList` is still the same, as T is derived.

compiler flags is use is `-fno-ms-compatibility -fno-delayed-template-parsing`.

I checked our code base and it looks like this is the only class template that it's arguments is a single `auto`, which is then used as a member of a struct/class

**Comment by martinbonaciugome** [comment]
ok and I have a smaller project that I've just tried, and MSVC 19.38.33129.0 works

**Comment by martinbonaciugome** [comment]
actaully i have a simple repro
```
template<auto>
struct A
{
  A()
  {
  }
  ~A()
  {
  }
};

struct B
{
  int X;
};

struct C
{
  int X;
};

struct D
{
  A<&B::X> B;
  A<&C::X> C;
};

int main(int, char* [])
{
  D d{};
  return 0;
}
```

**Comment by dwblaikie** [comment]
Thanks for the reduced test case!

Looks like this is a mangling bug in clang-cl (probably due to the relatively new `auto` in template parameter lists):

MSVC mangles the two functions as:
```
??0?$A@$MPQB@@H0A@@@QAE@XZ
??0?$A@$MPQC@@H0A@@@QAE@XZ
```
Whereas `clang-cl` mangles them both to:
```
??0?$A@$0A@@@QEAA@XZ
```
https://godbolt.org/z/E8n37jcG3 demonstrates the issue (comment out one of the declarations in `f1` to get the build to succeed with clang, then check the mangled name - comment out the other declaration and observe it mangles the same)

**Comment by MaxEW707** [comment]
I am interested in fixing this bug. Not sure how long it will take me so if anyone else is working on this in parallel and beats me to a PR, no worries.

I am slowly working my way creating local tests and reversing msvc's new name mangling scheme here.

**Comment by MaxEW707** [comment]
https://godbolt.org/z/oq657n9eK as reference for the upcoming statements.

VS2017 doesn't correctly mangle C++17 template auto. It erroneously emits duplicate mangled names for different deduced types.
This was fixed in VS2019.

We correctly mangle integral types including enums.
This is special cased inside [MicrosoftCXXNameMangler::mangleIntegerLiteral](https://github.com/llvm/llvm-project/blob/main/clang/lib/AST/MicrosoftMangle.cpp#L1557).

We don't correctly mangle pointer to member data, pointer to member function, pointer to data and pointer to function.

I also suspect float non type template parameters in C++20 are mangled incorrectly as well.

I am hoping to get a PR up for the various pointer fixes sometime in the next two weeks assuming nothing pops up :).

**Comment by MaxEW707** [comment]
Quick update since its been a bit over 2 weeks. I have this fixed locally with unit tests for function pointers, pointers, pointer to member functions and pointer to data members.

I got side tracked tracking down this mismangle, https://github.com/llvm/llvm-project/issues/92204, that I found when writing unit tests.
By end of week I'll get the PR up to fix this mismangle and then dig into #92204
